pulsar-broker/src/main/java/org/apache/pulsar/broker/delayed/BucketDelayedDeliveryTracker.java

```diff
@@ -0,0 +1,569 @@
+    }
+
+    private BucketState createBucket(long startLedgerId, long endLedgerId) {
+        BucketState bucketState = new BucketState(startLedgerId, endLedgerId);
+        immutableBuckets.put(Range.closed(startLedgerId, endLedgerId), bucketState);
+        return bucketState;
+    }
+
+    private CompletableFuture<Long> asyncSaveBucketSnapshot(
+            BucketState bucketState, SnapshotMetadata snapshotMetadata,
+            List<SnapshotSegment> bucketSnapshotSegments) {
+
+        return bucketSnapshotStorage.createBucketSnapshot(snapshotMetadata, bucketSnapshotSegments)
+                .thenCompose(newBucketId -> {
+                    bucketState.setBucketId(newBucketId);
+                    String bucketKey = bucketState.bucketKey();
+                    return putBucketKeyId(bucketKey, newBucketId).exceptionally(ex -> {
+                        log.warn("Failed to record bucketId to cursor property, bucketKey: {}", bucketKey);
+                        return null;
+                    }).thenApply(__ -> newBucketId);
+                });
+    }
+
+    private CompletableFuture<Void> putBucketKeyId(String bucketKey, Long bucketId) {
+        Objects.requireNonNull(bucketId);
+        return executeWithRetry(() -> cursor.putCursorProperty(bucketKey, String.valueOf(bucketId)),
+                ManagedLedgerException.BadVersionException.class);
+    }
+
+    private CompletableFuture<Long> asyncCreateBucketSnapshot() {
+        TripleLongPriorityQueue priorityQueue = super.getPriorityQueue();
+        if (priorityQueue.isEmpty()) {
+            return CompletableFuture.completedFuture(-1L);
+        }
+        long numMessages = 0;
+
+        final long startLedgerId = lastMutableBucketState.startLedgerId;
+        final long endLedgerId = lastMutableBucketState.endLedgerId;
+
+        List<SnapshotSegment> bucketSnapshotSegments = new ArrayList<>();
+        List<SnapshotSegmentMetadata> segmentMetadataList = new ArrayList<>();
+        Map<Long, RoaringBitmap> bitMap = new HashMap<>();
+        SnapshotSegment.Builder snapshotSegmentBuilder = SnapshotSegment.newBuilder();
+        SnapshotSegmentMetadata.Builder segmentMetadataBuilder = SnapshotSegmentMetadata.newBuilder();
+
+        long currentTimestampUpperLimit = 0;
+        while (!priorityQueue.isEmpty()) {
+            long timestamp = priorityQueue.peekN1();
+            if (currentTimestampUpperLimit == 0) {
+                currentTimestampUpperLimit = timestamp + timeStepPerBucketSnapshotSegment - 1;
+            }
+
+            long ledgerId = priorityQueue.peekN2();
+            long entryId = priorityQueue.peekN3();
+
+            checkArgument(ledgerId >= startLedgerId && ledgerId <= endLedgerId);
+
+            // Move first segment of bucket snapshot to sharedBucketPriorityQueue
+            if (segmentMetadataList.size() == 0) {
+                sharedBucketPriorityQueue.add(timestamp, ledgerId, entryId);
+            }
+
+            priorityQueue.pop();
```

Comment:
I'm not sure If we exist some race condition cause poped entry is not the peeked entry.
e.g. After the peek, another thread popped up with this entry.

Reply:
Same for https://github.com/apache/pulsar/pull/17611#discussion_r995249586